File: python/剑指offer/面试题32 - III. 从上到下打印二叉树 III.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue, res= [root],[root.val]
        i = 1
        while queue:
            temp = []
            for node in queue:
                if node.left: temp.append(node.left)
                if node.right: temp.append(node.right)
            queue = temp
            if temp:
                if i % 2 == 0:
                    res.append(temp[::-1])
                else:
                    res.append(temp)
            i+=1
        return res

File: go/剑指offer/面试题32 - III. 从上到下打印二叉树 III.go
package main

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	queue := []*TreeNode{root}
	var res [][]int
	var j int
	for len(queue) > 0 {
		l := len(queue)
		var tmp []int
		for i := 0; i < l; i++ {
			node := queue[0]
			tmp = append(tmp, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			queue = queue[1:]
		}
		if j%2 == 0 {
			res = append(res, tmp)
		} else {
			reverse(tmp)
			res = append(res, tmp)
		}
	}
	return res
}

func reverse(tmp []int) {
	for i, j := 0, len(tmp)-1; i < j; i, j = i+1, j-1 {
		tmp[i], tmp[j] = tmp[j], tmp[i]
	}
}
